datafusion/physical-expr/src/aggregate/average.rs

```diff
@@ -106,11 +100,29 @@ impl AggregateExpr for Avg {
     }
 
     fn create_accumulator(&self) -> Result<Box<dyn Accumulator>> {
-        Ok(Box::new(AvgAccumulator::try_new(
-            // avg is f64 or decimal
-            &self.sum_data_type,
-            &self.rt_data_type,
-        )?))
+        use DataType::*;
+        // instantiate specialized accumulator based for the type
+        match (&self.sum_data_type, &self.rt_data_type) {
+            (Float64, Float64) => {
+                Ok(Box::new(AvgAccumulator::new(self.pre_cast_to_sum_type)))
```

Comment:
This whole precast thing seems like a hack, imo this should be being handled by the type coercion machinery, not internal to the aggregator...